feat: secure, small, multi-stage Dockerfile

pfarikrispy opened the pull request:
* separate build into stage
* isolate base image and apt-get
* isolate the app build
* copy only the app into a secure Chainguard Python image that's updated frequently
* group stmnts for less layers
* use image tag SHAs for more (supply chain) security
* make non-root user explicit in final stage
* smaller image that's also more secure

pfarikrispy committed:
feat: secure, small, multi-stage Dockerfile

GDay commented on back/Dockerfile:
```diff
@@ -1,16 +1,27 @@
 
-CMD ["/usr/bin/supervisord", "-c", "/etc/supervisor/conf.d/supervisord.conf"]
+# checkov:skip=CKV_DOCKER_7: zero-CVE fresh image every build
+FROM cgr.dev/chainguard/python:latest@sha256:136aad7020e00a98f617f3d3343cc7601b7823405eb2bc581eae5f5a8c21e8d0
```
Shouldn't this be `3.13` instead of `latest`? It looks like `latest` will download python 3.14?

GDay reviewed (commented):
Thanks! First time hearing about chainguard, looks good. One question before I merge this.

pfarikrispy committed:
use nvidia distroless python 3.13 image

replace chainguard's latest python container (3.14) with a distroless build from nvidia that uses python 3.13

pfarikrispy committed:
Update multi-stage Dockerfile

redo layers:
- use debian 13 (trixie) to install supervisord
- use python 3.13 with trixie to build the app
- copy everything into the distroless python image from nvidia

pfarikrispy commented:
@GDay I'm sorry but I'm having troubles redesigning the Dockerfile so that it become a secure multi-stage file because I don't understand Python building enough. I'm expecting a `requirements.txt` somewhere or a [pyproject.toml](https://python-poetry.org/docs/pyproject/). I do find the Pipfile but I don't know how to deal with that.

I have tried several iterations of the layers locally but each time pipenv complains with the same error:
```
Dockerfile:17
--------------------
  15 |     COPY ./back/ /app
  16 |     WORKDIR /app
  17 | >>> RUN pipenv install --system
  18 |     RUN django-admin compilemessages 
  19 |     
--------------------
ERROR: failed to build: failed to solve: process "/bin/sh -c pipenv install --system" did not complete successfully: exit code: 2
```

I'd love to help you get an improved container spec but I need some assistance with Python building.

I also tried to just use one layer to build the app and then copy artifacts into the distroless container, but it's not building the app (yet).
```
# syntax=docker/dockerfile:1
# python images are based off of Debian
FROM python:3.13-slim-trixie AS app
ENV PYTHONUNBUFFERED=1
USER root
RUN mkdir /var/run/supervisord && mkdir /var/log/supervisord
# trunk-ignore(hadolint/DL3008)
RUN apt-get -y update && apt-get install -y --no-install-recommends \
  gettext \
  supervisor \
  && rm -rf /var/lib/apt/lists/*
# trunk-ignore(hadolint/DL3013)
RUN python -m pip install --no-cache-dir pipenv
RUN mkdir -p /app
COPY ./back/ /app
WORKDIR /app
RUN pipenv install --system
RUN django-admin compilemessages 

# checkov:skip=CKV_DOCKER_7: zero-CVE fresh image every build
FROM nvcr.io/nvidia/distroless/python:3.13-v3.1.1
USER root
HEALTHCHECK NONE
WORKDIR /
COPY ./supervisord.conf /etc/supervisor/conf.d/supervisord.conf
COPY --from=app /var/log/supervisord /var/run/supervisord ./
COPY --from=app /usr/bin/supervisord /usr/bin/supervisord
COPY --from=app /app /app
USER 65532
CMD ["/usr/bin/supervisord", "-c", "/etc/supervisor/conf.d/supervisord.conf"]
```

Do you have any ideas to make that work?

pfarikrispy committed:
Merge branch 'master' into multistage
